[Gutenberg] Emit editor save events

SiobhyB opened the pull request:
Fixes https://github.com/wordpress-mobile/gutenberg-mobile/issues/5666

## Related PRs

* `Gutenberg`: https://github.com/WordPress/gutenberg/pull/49781
* `Gutenberg Mobile`: https://github.com/wordpress-mobile/gutenberg-mobile/pull/5663
* `Jetpack`: https://github.com/Automattic/jetpack/pull/30131
* `Android`: https://github.com/wordpress-mobile/WordPress-Android/pull/18279

## To Test

After making some edits to a post or page, there are multiple flows to manually save based on whether a post/page is a draft, published, or scheduled. These have been outlined below.

<details>
 <summary>Flows for saving a draft ⤵️</summary>

* Tap the **Update** button.
* Tap the X icon to exit the post and **Update Draft**.
* Tap the three dots to the upper right and then **Publish**.
* The the three dots to the upper right to get to Post Settings. From here, set the post's publish date and then **Schedule**.

</details>

<details>
 <summary>Flows for saving a published post ⤵️</summary>

* Tap the **Update** button.
* Tap the X icon to exit the post and **Update Post**.
* The the three dots to the upper right to get to Post Settings. From here, set the post's publish date and then **Schedule**.

</details>

<details>
 <summary>Flows for saving a scheduled post ⤵️</summary>

* Tap the **Update** button.
* Tap the X icon to exit the post and **Update Post**.
* The the three dots to the upper right to get to Post Settings. From here, set the post's publish date and then **Schedule**.

</details>

## Regression Notes
1. Potential unintended areas of impact

As this PR is only emitting a save event, and not changing any functionality, I can't foresee any unintended impact.

2. What I did to test those areas of impact (or what existing automated tests I relied on)

Although I don't believe there will be unintended impact, I did some quick manual smoke testing of the editor's functionality and saving as part of this PR.

3. What automated tests I added (or what prevented me from doing so)

N/A

<hr />

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

SiobhyB committed:
chore: Update Gutenberg Mobile reference

SiobhyB committed:
feat: Add function for emitting save event

SiobhyB committed:
feat: Emit save event after manual save

SiobhyB committed:
Merge branch 'trunk' into gutenberg/emit-editor-save-events

SiobhyB committed:
chore: Update Gutenberg Mobile reference

fluiddot reviewed (commented):
@SiobhyB I can't build the app locally to test these changes, I'm getting the same failure as [in the CI jobs](https://buildkite.com/automattic/wordpress-ios/builds/13761#01879157-565d-47d7-9726-3f09b7be5b86/6-5489) when building the project. Would you mind taking a look?

SiobhyB committed:
refactor: Remove accidental deletion of lines

SiobhyB committed:
refactor: Remove second function call

SiobhyB commented:
@fluiddot, whoops, sorry about that, that should be fixed as of https://github.com/wordpress-mobile/WordPress-iOS/pull/20522/commits/d7584cea3350e88e1898b5b0e8f25d1462f4c95b (there was a second function call, which I overlooked as it had been removed in my local set up 🤦‍♀️).

fluiddot commented on WordPress/Classes/ViewRelated/Post/PostEditor+Publish.swift:
```diff
@@ -57,6 +59,10 @@ extension PublishingEditor {
 
     }
 
+    func emitPostSaveEvent() {
+
+    }
+
```
I tried to remove this function to check if it's needed, and got compilation errors in `AztecPostViewController` and `StoryEditor`. I think it's ok to define the default behavior here as it's optional, but I'm wondering if we could add a comment about why this function is empty, WDYT?

SiobhyB replied:
Sure thing! I followed the same pattern that I saw for other optional functions like this ([example](https://github.com/wordpress-mobile/WordPress-iOS/blob/d7584cea3350e88e1898b5b0e8f25d1462f4c95b/WordPress/Classes/ViewRelated/Post/PostEditor%2BPublish.swift#L58-L60)), which don't have comments, but agree it's a good idea to add for clarification.

SiobhyB replied:
Added in https://github.com/wordpress-mobile/WordPress-iOS/pull/20522/commits/07ddf84736968bc5bd9f782088186f68323a499a, I took the liberty of also adding the same comment to the other empty function in the file, for consistency.

fluiddot commented on WordPress/Classes/ViewRelated/Post/PostEditor+Publish.swift:
```diff
@@ -95,6 +101,8 @@ extension PublishingEditor {
 
         mapUIContentToPostAndSave(immediate: true)
 
+        emitPostSaveEvent()
```
I noticed that some lines below the publishing can be canceled, so maybe we should call this function once we are sure that the post will be saved, WDYT?

https://github.com/wordpress-mobile/WordPress-iOS/blob/d7584cea3350e88e1898b5b0e8f25d1462f4c95b/WordPress/Classes/ViewRelated/Post/PostEditor%2BPublish.swift#L106-L110

I potential place would be right before we track the `PostSave` event:
https://github.com/wordpress-mobile/WordPress-iOS/blob/d7584cea3350e88e1898b5b0e8f25d1462f4c95b/WordPress/Classes/ViewRelated/Post/PostEditor%2BPublish.swift#L151-L167

SiobhyB replied:
That logic makes sense by me! Updated in https://github.com/wordpress-mobile/WordPress-iOS/pull/20522/commits/4e944ebdd90613e7a0960d37f64f5c4afe5bbb92.

SiobhyB committed:
docs: Clarifying comments within empty funcs

SiobhyB committed:
chore: Update Gutenberg Mobile reference

SiobhyB committed:
refactor: Change location of emitPostSaveEvent()

Feedback ref: https://github.com/wordpress-mobile/WordPress-iOS/pull/20522#discussion_r1169930544

fluiddot commented on Podfile.lock:
```diff
@@ -4,21 +4,21 @@ PODS:
+  - AppAuth/Core (1.6.1)
+  - AppAuth/ExternalUserAgent (1.6.1):
     - AppAuth/Core
-  - AppCenter (4.4.1):
-    - AppCenter/Analytics (= 4.4.1)
-    - AppCenter/Crashes (= 4.4.1)
-  - AppCenter/Analytics (4.4.1):
+  - AppCenter (4.4.3):
+    - AppCenter/Analytics (= 4.4.3)
+    - AppCenter/Crashes (= 4.4.3)
+  - AppCenter/Analytics (4.4.3):
     - AppCenter/Core
-  - AppCenter/Core (4.4.1)
-  - AppCenter/Crashes (4.4.1):
+  - AppCenter/Core (4.4.3)
+  - AppCenter/Crashes (4.4.3):
     - AppCenter/Core
-  - AppCenter/Distribute (4.4.1):
+  - AppCenter/Distribute (4.4.3):
     - AppCenter/Core
   - Automattic-Tracks-iOS (0.13.0):
     - Sentry (~> 7.25)
```
@SiobhyB I noticed that several pods have been updated. However, I don't think this is expected when only bumping the Gutenberg Mobile reference. Usually, only the pods related to Gutenberg are the ones that are updated 🤔.

fluiddot replied:
Not sure if it's caused by this, but I'm getting the following error when building the app:
```
./WordPress-iOS/WordPress/Classes/Utility/Kanvas/KanvasCameraAnalyticsHandler.swift:5:20: error build: Type 'KanvasAnalyticsHandler' does not conform to protocol 'KanvasAnalyticsProvider'
```

SiobhyB replied:
Hmmm, I'm not sure how those got updated along the way 🤔 this should be addressed in the most recent commits, though.

fluiddot replied:
Great, I've just checked the last changes and it looks good. Thanks!

SiobhyB committed:
Merge branch 'trunk' into gutenberg/emit-editor-save-events

SiobhyB committed:
Revert "chore: Update Gutenberg Mobile reference"

This reverts commit aa93c70da76beb1427325048a522e4e995ae0f87.

SiobhyB committed:
chore: Update Gutenberg Mobile reference

SiobhyB committed:
chore: Update Gutenberg Mobile reference

fluiddot reviewed (approved):
Approved via https://github.com/Automattic/jetpack/pull/30131#pullrequestreview-1392487317.

**NOTE:** I've only tested these changes building the app locally.

fluiddot commented on Podfile:
```diff
@@ -91,7 +91,7 @@ def shared_style_pods
 
 def gutenberg_pods
-  gutenberg tag: 'v1.93.0'
+  gutenberg commit: 'a9bafbb3e6ecf1b0fd2532af5a72a5756f5d1690'
```
Heads up that the Gutenberg Mobile reference doesn't point to the last commit from https://github.com/wordpress-mobile/gutenberg-mobile/pull/5663: [`3b7427e1f12a6e34f5b5760c29a3db91cca33857`](https://github.com/wordpress-mobile/gutenberg-mobile/pull/5663/commits/3b7427e1f12a6e34f5b5760c29a3db91cca33857)

SiobhyB committed:
Merge branch 'trunk' into gutenberg/emit-editor-save-events

SiobhyB committed:
Update Gutenberg Mobile reference

SiobhyB committed:
Fix incorrect checksum
